scripts/execute_code.py

```diff
@@ -86,3 +86,21 @@ def execute_shell(command_line):
     os.chdir(current_dir)
 
     return output
+def execute_shell_popen(command_line):
+
+    current_dir = os.getcwd()
+
+    if not WORKSPACE_FOLDER in current_dir: # Change dir into workspace if necessary
+        work_dir = os.path.join(os.getcwd(), WORKSPACE_FOLDER)
+        os.chdir(work_dir)
+
+    print (f"Executing command popen '{command_line}' in working directory '{os.getcwd()}'")
+
+    do_not_show_output = subprocess.DEVNULL
+    process = subprocess.Popen(command_line, shell=True, stdout=do_not_show_output, stderr=do_not_show_output)
```

Comment:
Seems good but how will AutoGPT ever close/kill the process?  How does it know what processes are running in the background?

Reply:
Just below the code you linked is: `return f"Subprocess started with PID:'{str(process.pid)}'"` 

This displays as `SYSTEM:  Command execute_shell_popen returned: Subprocess started with PID:'14424'` which it can see. I saw it save this PID to a file and kill it when asked.